draft-specifications/Addresses-References.md

```diff
@@ -0,0 +1,222 @@
+the sender must establish a session with a protocol that both the sender
+and receiver support.
+
+> For example, if the sender node is a web page and the reference on a web
+> server, these nodes might both be able to connect using version 1 of
+> OCapN and rely on WebSocket for message framing and HTTP with TLS and a
+> Certificate Authority for transport.
+> The server might also provide addresses for that reference for TLS with a
+> Diffie-Helman key exchange, Netstring for message framing, and OCapN versions 1
+> and 2 for CapTP.
+
+# Network Layers
+
+OCapN specifies the names of multiple addressable network layers.
+A node may have multiple network layers and a reference may have multiple
+addresses and multiple addresses may share a network layer.
+
+## TCP with Syrup, Version 0
+
+For testing purposes only, the network layer named `tcp-syr0` combines:
+
+* TCP for sessions,
+* Netstring for message framing,
+* Syrup OCapN Version 0 (draft)
+
+> We advise the TCP transport to listen specifically on IP `127.0.0.1` (rather
+> than the typical default of `0.0.0.0`) to make the node unreachable from
+> other hosts and avoid use on multi-tenant hosts.
+> For private connections between a user and a user-agent on a multi-tenant
+> system, relay OCapN through a UNIX domain socket or a Windows named pipe.
+
+The Syrup serialization is a record with the symbol `'tcp-syr0` followed
+by a TCP address, including the port.
+This network layer does not have a default port.
+
+```
+<'tcp-syr0 host> ; string
+```
+
+The JSON serialization is an object where:
+
+* The `protocol` is `tcp-syr0`.
+* The `host` is the TCP address of the node, including the port.
+
+The URL serialization:
+
+* The `protocol` is `tcp-syr0`.
+* The `host` is the TCP address of the node.
+
+> For example, `tcp-syr0://127.0.0.1:32768` is the address of a node reachable
+> locally on port 32768.
+
+## Tor with Syrup, Version 0
+
+The network layer named `tor-syr0` combines:
+
+* Tor for sessions,
+* Netstring for message framing, and
+* Syrup OCapN Version 0 (draft)
+
+The Syrup serialization is a record with the symbol `'tor-syr0` followed by an
+onion address (as a string), including the `.onion` virtual top-level-domain.
+
+```
+<'tor-syr0 onion> ; string
+```
+
+The JSON serialization is an object:
+
+* The `protocol` is `tor-syr0`.
+* The `address` is the string onion address, including the `.onion` virtual
+  top-level-domain.
+
+The URL serialization:
+
+* The `protocol` is `tor-syr0`.
+* The host is the Tor onion address, including the `.onion` virtual
+  top-level-domain.
+
+> For example, `tor-syr0://b0b5c0ffeefacade.onion` is the address of an OCapN
+> node reachable over Tor with the hidden service onion address
+> `b0b5c0ffeefacade.onion`.
+
+## TLS WebSocket with Syrup, Version 0
+
+The network layer named `wss-syr0` combines:
+
+* Signed Certificate TLS for sessions,
+* HTTP for transport,
+* WebSocket for message framing, and
+* Syrup OCapN Version 0 (draft)
+
```

Comment:
Anything touching TLS should allow certificate pinning to facilitate using encryption without touching PKI.

The cert chain is a hairy thing on its own, but at least the basic "pin the root" has great feature-to-weight ratio.

Reply:
This is not in my wheelhouse. I’d welcome suggestion text!

Reply:
Basically there should be an array of CN/SHA256 pairs. In the simplest case of self-signed certificate it would be just one element with CN=hostname and hash of the certificate.

Something like `pin=hostname/NWZiNGJhMWE2NTFiYWU4MDU3ZWM2YjVjZGFmYzkzZmE3ZTBiN2Q5NDRkNmYwMmE0Yjc1MWRlNGUxNTQ2NGRlZiAqLQo=&pin=CA%20Root/M2Y0MGYxNTg4NjczMzcyODc0ZTVjOTI2NDc5NGE3NWM2ZmVhYzFlNDExNTI5YzcxMTEwYjZmOGNlZjlhMTdlOCAqLQo=` in URI form.
In JSON:
```json
{ ...
, "pins":
  [ {"cn": "hostname", "sha256": "NWZiNGJhMWE2NTFiYWU4MDU3ZWM2YjVjZGFmYzkzZmE3ZTBiN2Q5NDRkNmYwMmE0Yjc1MWRlNGUxNTQ2NGRlZiAqLQo="}
  , {"cn": "CA Root", "sha256": "M2Y0MGYxNTg4NjczMzcyODc0ZTVjOTI2NDc5NGE3NWM2ZmVhYzFlNDExNTI5YzcxMTEwYjZmOGNlZjlhMTdlOCAqLQo="}
  ]
}
```